### PR TITLE
Add diff write mode https://github.com/nrc/rustfmt/issues/261

### PR DESCRIPTION
**Pull request by sinhpham**

### CODE REVIEW

**Comment by sinhpham** [comment]
We should probably change the way we display mismatches in test to reuse the print_diff function (showing 1 file name with multiple diff sections instead of showing file name in every diff section).

**Comment by marcusklaas** [comment]
Nice! 👌

Could you move the diff stuff to its own module? That way we'll keep the root namespace (relatively) clean.

-----Original Message-----

**Comment by marcusklaas** [comment]
It looks like this will only show the diff for the root file? I am not sure if we're currently doing this correctly for tests though..

-----Original Message-----

**Comment by marcusklaas** [comment]
Also, we should probably update this string literal https://github.com/nrc/rustfmt/blob/master/src/bin/rustfmt.rs#L85

File: src/utils.rs [hunk 1]
```diff
@@ -202,6 +206,109 @@ pub fn wrap_str<S: AsRef<str>>(s: S, max_width: usize, width: usize, offset: usi
...
+    results.remove(0);
+
+    results
+}
+
+pub fn print_diff(diff: Vec<Mismatch>) {
+    let mut t = term::stdout().unwrap();
+    for mismatch in diff {
+        t.fg(term::color::BRIGHT_WHITE).unwrap();
+        writeln!(t, "\nDiff at {}:", mismatch.line_number).unwrap();
```
Comment (marcusklaas) [inline src/utils.rs]: nit: we should probably clarify that the number is a line number, e.g. diff at line 30

**Comment by marcusklaas** [comment]
And yes, reusing `print_diff` is a good idea. I'm not convinced we should strip the filename from every "mismatch", though. Repeating them for every mismatch makes it easier to orientate oneself, especially for large diffs on many files.

File: src/filemap.rs [hunk 2]
```diff
@@ -104,6 +105,17 @@ fn write_file(text: &StringBuffer,
...
         }
+        WriteMode::Diff => {
+            println!("Diff of {}:\n", filename);
+            let mut f = try!(File::open(filename));
+            let mut ori_string = String::new();
+            try!(f.read_to_string(&mut ori_string));
+            let mut v = Vec::new();
+            try!(write_system_newlines(&mut v, text, config));
+            let text_str = String::from_utf8(v).unwrap();
+            let diff = make_diff(&ori_string, &text_str, 3);
```
Comment (marcusklaas) [inline src/filemap.rs]: The order of original, formatted text is inverted compared to tests here. I don't know what the more natural choice is, but we should at least be consistent.

Comment (sinhpham) [inline src/filemap.rs]: I think this order is more natural, at least for the diff mode. It reports the old code as deleted in red and the new formatted code as added in green.

**Comment by nrc** [comment]
@sinhpham for future reference, you get better commit history if you use `git rebase` rather than `git merge`. Could you change the commit message of the last commit please? (Or just merge it with previous commit).

**Comment by sinhpham** [comment]
@nrc noted and I've updated the PR. I'm not use to the rebasing work flow.

**Comment by marcusklaas** [comment]
Excellent!

**Pull request merged**